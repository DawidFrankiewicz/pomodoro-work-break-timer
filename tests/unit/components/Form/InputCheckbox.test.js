import { mount } from '@vue/test-utils';

const factory = async (propsData) => {
    const component = await import('@/components/Form/InputCheckbox.vue');

    return mount(component.default, {
        global: {
            stubs: ['FontAwesomeIcon'],
        },
        props: propsData,
    });
};

describe('InputCheckbox.test.js', () => {
    test('Should render correctly', async () => {
        const component = await import('@/components/Form/InputCheckbox.vue');
        expect(component).toBeDefined();

        const wrapper = await factory({
            value: true,
            label: 'Test',
        });

        // Check if component matches snapshot (HTML structure)
        // https://vitest.dev/guide/snapshot#updating-snapshots
        expect(wrapper.html()).toMatchSnapshot();
        // Check if all important elements are rendered
        expect(wrapper.find('[data-e2e="label"]').exists()).toBe(true);
        expect(wrapper.find('[data-e2e="input"]').exists()).toBe(true);
        expect(wrapper.find('[data-e2e="checkbox"]').exists()).toBe(true);
    });
});

describe.concurrent(
    'Should have input value equal to passed prop',
    async () => {
        test.each([
            [true, true],
            [false, false],
        ])('Props: Value %s', async (input, expected) => {
            const wrapper = await factory({
                value: input,
            });

            expect(
                wrapper.find('[data-e2e="input"]').wrapperElement.value
            ).toBe(String(expected));
        });
    }
);

describe.concurrent('Should change value to opposite', async () => {
    test.each([
        [false, true, 0],
        [true, false, 1],
    ])(`From %s to %s`, async (input, expected) => {
        const wrapper = await factory({
            value: input,
        });

        // Check displayed initial value
        expect(wrapper.find('[data-e2e="input"]').wrapperElement.value).toBe(
            String(input)
        );

        // Click checkbox
        await wrapper.find('[data-e2e="checkbox"]').trigger('click');
        await wrapper.setProps({
            value: wrapper.emitted('update')[0][0],
        });

        // Check if end value is correct
        // Check displayed value
        expect(wrapper.find('[data-e2e="input"]').wrapperElement.value).toBe(
            String(expected)
        );
        // Check last emitted value
        expect(wrapper.emitted('update')[0][0]).toBe(expected);
    });
});

describe.concurrent('Should display label correctly', async () => {
    test('Label: "Test"', async () => {
        const wrapper = await factory({
            label: 'Test',
            value: false,
        });

        expect(wrapper.find('[data-e2e="label"]').exists()).toBe(true);
        expect(wrapper.find('[data-e2e="label"]').text()).toBe('Test');
    });

    test('Label: null', async () => {
        const wrapper = await factory({
            label: null,
            value: false,
        });

        expect(wrapper.find('[data-e2e="label"]').exists()).toBe(false);
    });
});
