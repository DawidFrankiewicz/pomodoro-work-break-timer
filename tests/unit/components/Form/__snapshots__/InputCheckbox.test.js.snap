// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputCheckbox.test.js > Should render correctly 1`] = `
"<label><span data-e2e="label" class="block font-bold">Test</span>
  <div class="flex h-8"><input data-e2e="input" class="absolute appearance-none" type="checkbox" value="true"><button data-e2e="checkbox" class="h-8 w-16 rounded-md transition-all duration-75 disabled:opacity-40 ml-[2px] mt-[2px] border-b-2 border-r-2 border-emerald-400 bg-emerald-100 hover:bg-emerald-50">On <font-awesome-icon-stub icon="fas,check"></font-awesome-icon-stub></button></div>
</label>"
`;
